File: app/src/main/java/ru/iandreyshev/parserrss/models/async/GetRssFromNetTask.java
package ru.iandreyshev.parserrss.models.async;

import javax.annotation.Nullable;

import ru.iandreyshev.parserrss.models.rss.ViewRss;
import ru.iandreyshev.parserrss.models.rss.RssParser;
import ru.iandreyshev.parserrss.models.repository.Rss;
import ru.iandreyshev.parserrss.models.web.HttpRequestHandler;
import ru.iandreyshev.parserrss.models.web.IHttpRequestResult;

abstract class GetRssFromNetTask extends Task<String, Void, ViewRss> {
    private final HttpRequestHandler mRequestHandler;
    private final IEventListener mListener;
    private Rss mNewRss;

    protected interface IEventListener extends ITaskListener<ViewRss> {
        void onInvalidUrl();

        void onNetError(final IHttpRequestResult requestResult);

        void onParserError();

        void onSuccess(final ViewRss result);
    }

    GetRssFromNetTask(final IEventListener listener, String url) {
        super(listener);
        mListener = listener;
        mRequestHandler = new HttpRequestHandler(url);
    }

    @Nullable
    protected abstract Rss onSuccess(final Rss rss);

    boolean isUrlValid() {
        if (mRequestHandler.getState() != HttpRequestHandler.State.BadUrl) {
            return true;
        }

        setResultEvent(mListener::onInvalidUrl);

        return false;
    }

    boolean getRssFromNet() {
        mRequestHandler.sendGet();

        if (mRequestHandler.getState() == HttpRequestHandler.State.Success) {
            return true;
        }

        setResultEvent(() -> mListener.onNetError(mRequestHandler));

        return false;
    }

    boolean parseRss() {
        if ((mNewRss = RssParser.parse(mRequestHandler.getResponseBodyAsString())) != null) {
            mNewRss.setUrl(mRequestHandler.getUrlStr());

            return true;
        }

        setResultEvent(mListener::onParserError);

        return false;
    }

    @Nullable
    @Override
    protected final ViewRss doInBackground(final String... strings) {
        if (!isUrlValid() || !getRssFromNet() || !parseRss()) {
            return null;
        }

        final Rss rss = onSuccess(mNewRss);

        return rss == null ? null : new ViewRss(rss);
    }

    @Nullable
    protected final String getUrl() {
        return mRequestHandler.getUrlStr();
    }
}

File: app/src/main/kotlin/ru/iandreyshev/parserrss/models/async/GetRssFromNetTask.kt
package ru.iandreyshev.parserrss.models.async

import ru.iandreyshev.parserrss.models.repository.Rss
import ru.iandreyshev.parserrss.models.rss.ParserEngine
import ru.iandreyshev.parserrss.models.web.HttpRequestHandler
import ru.iandreyshev.parserrss.models.web.IHttpRequestResult

abstract class GetRssFromNetTask(
        private val mListener: IEventListener,
        url: String) : Task<String, Void, Rss>(mListener) {

    companion object {
        private const val MAX_ARTICLES_COUNT = 64
        private const val MAX_CONNECTION_TIMEOUT_SEC = 4L
        private const val MAX_READ_TIMEOUT_SEC = 4L
        private const val MAX_WRITE_TIMEOUT_SEC = 4L
    }

    private var requestHandler: HttpRequestHandler = HttpRequestHandler(url)
    private var resultRss: Rss? = null

    open fun isUrlValid(): Boolean {
        return when (requestHandler.state != HttpRequestHandler.State.BAD_URL) {
            true -> true
            false -> {
                setResultEvent { mListener.onInvalidUrl() }
                false
            }
        }
    }

    open fun getRssFromNet(): Boolean {
        requestHandler.connectionTimeoutSec = MAX_CONNECTION_TIMEOUT_SEC
        requestHandler.readTimeoutSec = MAX_READ_TIMEOUT_SEC
        requestHandler.writeTimeoutSec = MAX_WRITE_TIMEOUT_SEC
        requestHandler.sendGet()

        return when (requestHandler.state == HttpRequestHandler.State.SUCCESS) {
            true -> true
            false -> {
                setResultEvent { mListener.onNetError(requestHandler) }
                false
            }
        }
    }

    open fun parseRss(): Boolean {
        val rss = ParserEngine.parse(requestHandler.bodyAsString, MAX_ARTICLES_COUNT)

        return when (rss) {
            null -> {
                setResultEvent { mListener.onParserError() }
                false
            }
            else -> {
                rss.url = requestHandler.urlString
                resultRss = rss
                true
            }
        }
    }

    override fun doInBackground(vararg args: String): Rss? {
        if (!isUrlValid() || !getRssFromNet() || !parseRss()) {
            return null
        }

        onSuccess(resultRss ?: return null)

        return resultRss
    }

    protected abstract fun onSuccess(rss: Rss)

    interface IEventListener : ITaskListener<String, Void, Rss> {
        fun onInvalidUrl()

        fun onNetError(requestResult: IHttpRequestResult)

        fun onParserError()
    }
}
